File: backend/services/conversation_history_service.py
"""
Serviço de histórico de conversas salvas
"""
import json
from typing import List, Dict, Optional, Any
from datetime import datetime
from loguru import logger

from backend.database.database import Database


class ConversationHistoryService:
    """Gerencia histórico de conversas salvas"""
    
    def __init__(self, database: Database):
        """
        Inicializa o serviço de histórico
        
        Args:
            database: Instância do banco de dados
        """
        self.db = database
        logger.info("ConversationHistoryService inicializado")
    
    async def save_conversation(
        self,
        session_id: str,
        title: str,
        messages: List[Dict[str, str]],
        user_id: Optional[str] = None
    ) -> int:
        """
        Salva uma conversa no histórico
        
        Args:
            session_id: ID da sessão
            title: Título da conversa
            messages: Lista de mensagens [{role, content}]
            user_id: ID do usuário (opcional)
            
        Returns:
            ID da conversa salva
            
        Raises:
            ValueError: Se session_id ou title estiverem vazios
        """
        if not session_id or not title:
            raise ValueError("session_id e title são obrigatórios")
        
        if not messages:
            raise ValueError("messages não pode estar vazio")
        
        await self.db.connect()
        
        try:
            messages_json = json.dumps(messages, ensure_ascii=False)
            
            # Verifica se já existe conversa para esta sessão
            async with self.db._connection.execute(
                "SELECT id FROM saved_conversations WHERE session_id = ?",
                (session_id,)
            ) as cursor:
                existing = await cursor.fetchone()
            
            if existing:
                # Atualiza conversa existente
                await self.db._connection.execute(
                    """
                    UPDATE saved_conversations
                    SET title = ?, messages = ?, updated_at = CURRENT_TIMESTAMP, saved = 1
                    WHERE session_id = ?
                    """,
                    (title, messages_json, session_id)
                )
                conversation_id = existing[0]
                logger.info(f"Conversa atualizada: {conversation_id} (session: {session_id})")
            else:
                # Cria nova conversa
                cursor = await self.db._connection.execute(
                    """
                    INSERT INTO saved_conversations 
                    (session_id, title, messages, saved, user_id)
                    VALUES (?, ?, ?, 1, ?)
                    """,
                    (session_id, title, messages_json, user_id)
                )
                conversation_id = cursor.lastrowid
                logger.info(f"Conversa salva: {conversation_id} (session: {session_id})")
            
            await self.db._connection.commit()
            return conversation_id
            
        except Exception as e:
            await self.db._connection.rollback()
            logger.error(f"Erro ao salvar conversa: {e}")
            raise
    
    async def get_saved_conversations(
        self,
        limit: int = 50,
        offset: int = 0,
        user_id: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Lista conversas salvas
        
        Args:
            limit: Número máximo de resultados
            offset: Offset para paginação
            user_id: Filtrar por usuário (opcional)
            
        Returns:
            Lista de conversas salvas
        """
        await self.db.connect()
        
        try:
            if user_id:
                async with self.db._connection.execute(
                    """
                    SELECT id, session_id, title, created_at, updated_at
                    FROM saved_conversations
                    WHERE saved = 1 AND user_id = ?
                    ORDER BY created_at DESC
                    LIMIT ? OFFSET ?
                    """,
                    (user_id, limit, offset)
                ) as cursor:
                    rows = await cursor.fetchall()
            else:
                async with self.db._connection.execute(
                    """
                    SELECT id, session_id, title, created_at, updated_at
                    FROM saved_conversations
                    WHERE saved = 1
                    ORDER BY created_at DESC
                    LIMIT ? OFFSET ?
                    """,
                    (limit, offset)
                ) as cursor:
                    rows = await cursor.fetchall()
            
            conversations = []
            for row in rows:
                conversations.append({
                    "id": row[0],
                    "session_id": row[1],
                    "title": row[2],
                    "created_at": row[3],
                    "updated_at": row[4]
                })
            
            return conversations
            
        except Exception as e:
            logger.error(f"Erro ao listar conversas: {e}")
            raise
    
    async def get_conversation_by_id(self, conversation_id: int) -> Optional[Dict[str, Any]]:
        """
        Recupera uma conversa específica
        
        Args:
            conversation_id: ID da conversa
            
        Returns:
            Conversa completa ou None se não encontrada
        """
        await self.db.connect()
        
        try:
            async with self.db._connection.execute(
                """
                SELECT id, session_id, title, messages, created_at, updated_at, user_id
                FROM saved_conversations
                WHERE id = ? AND saved = 1
                """,
                (conversation_id,)
            ) as cursor:
                row = await cursor.fetchone()
            
            if not row:
                return None
            
            messages = json.loads(row[3]) if row[3] else []
            
            return {
                "id": row[0],
                "session_id": row[1],
                "title": row[2],
                "messages": messages,
                "created_at": row[4],
                "updated_at": row[5],
                "user_id": row[6]
            }
            
        except Exception as e:
            logger.error(f"Erro ao recuperar conversa: {e}")
            raise
    
    async def delete_conversation(self, conversation_id: int) -> bool:
        """
        Remove uma conversa salva
        
        Args:
            conversation_id: ID da conversa
            
        Returns:
            True se deletada, False se não encontrada
        """
        await self.db.connect()
        
        try:
            cursor = await self.db._connection.execute(
                "DELETE FROM saved_conversations WHERE id = ?",
                (conversation_id,)
            )
            await self.db._connection.commit()
            
            deleted = cursor.rowcount > 0
            if deleted:
                logger.info(f"Conversa deletada: {conversation_id}")
            
            return deleted
            
        except Exception as e:
            await self.db._connection.rollback()
            logger.error(f"Erro ao deletar conversa: {e}")
            raise
    
    async def update_conversation_title(
        self,
        conversation_id: int,
        new_title: str
    ) -> bool:
        """
        Atualiza o título de uma conversa
        
        Args:
            conversation_id: ID da conversa
            new_title: Novo título
            
        Returns:
            True se atualizada, False se não encontrada
        """
        if not new_title:
            raise ValueError("new_title não pode estar vazio")
        
        await self.db.connect()
        
        try:
            cursor = await self.db._connection.execute(
                """
                UPDATE saved_conversations
                SET title = ?, updated_at = CURRENT_TIMESTAMP
                WHERE id = ?
                """,
                (new_title, conversation_id)
            )
            await self.db._connection.commit()
            
            updated = cursor.rowcount > 0
            if updated:
                logger.info(f"Título da conversa atualizado: {conversation_id}")
            
            return updated
            
        except Exception as e:
            await self.db._connection.rollback()
            logger.error(f"Erro ao atualizar título: {e}")
            raise

File: mobile_app/lib/services/conversation_history_service.dart
/// Serviço para gerenciar histórico de conversas
import 'package:flutter/foundation.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/env.dart';
import '../models/conversation_history.dart';

class ConversationHistoryService {
  static const String baseUrl = '${Env.backendUrl}/api/conversations';

  /// Salva uma conversa
  Future<int> saveConversation({
    required String sessionId,
    required String title,
    String? userId,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/save'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'session_id': sessionId,
          'title': title,
          'user_id': userId,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        if (data['success'] == true) {
          return data['conversation_id'] as int;
        } else {
          throw Exception('Falha ao salvar conversa: ${data['detail']}');
        }
      } else if (response.statusCode == 404) {
        throw Exception('Sessão não encontrada. Envie uma mensagem primeiro.');
      } else {
        final error = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(error['detail'] ?? 'Erro ao salvar conversa');
      }
    } catch (e) {
      debugPrint('❌ Erro ao salvar conversa: $e');
      rethrow;
    }
  }

  /// Lista conversas salvas
  Future<List<ConversationSummary>> getConversations({
    int limit = 50,
    int offset = 0,
    String? userId,
  }) async {
    try {
      final queryParams = <String, String>{
        'limit': limit.toString(),
        'offset': offset.toString(),
      };
      if (userId != null) {
        queryParams['user_id'] = userId;
      }

      final uri = Uri.parse(baseUrl).replace(queryParameters: queryParams);
      final response = await http.get(uri);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        if (data['success'] == true) {
          final conversationsJson = data['conversations'] as List<dynamic>;
          return conversationsJson
              .map((json) => ConversationSummary.fromJson(
                  json as Map<String, dynamic>))
              .toList();
        } else {
          throw Exception('Falha ao listar conversas');
        }
      } else {
        final error = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(error['detail'] ?? 'Erro ao listar conversas');
      }
    } catch (e) {
      debugPrint('❌ Erro ao listar conversas: $e');
      rethrow;
    }
  }

  /// Recupera uma conversa específica
  Future<ConversationHistory> getConversation(int conversationId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/$conversationId'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        if (data['success'] == true) {
          return ConversationHistory.fromJson(
              data['conversation'] as Map<String, dynamic>);
        } else {
          throw Exception('Falha ao recuperar conversa');
        }
      } else if (response.statusCode == 404) {
        throw Exception('Conversa não encontrada');
      } else {
        final error = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(error['detail'] ?? 'Erro ao recuperar conversa');
      }
    } catch (e) {
      debugPrint('❌ Erro ao recuperar conversa: $e');
      rethrow;
    }
  }

  /// Deleta uma conversa
  Future<void> deleteConversation(int conversationId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/$conversationId'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        if (data['success'] != true) {
          throw Exception('Falha ao deletar conversa');
        }
      } else if (response.statusCode == 404) {
        throw Exception('Conversa não encontrada');
      } else {
        final error = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(error['detail'] ?? 'Erro ao deletar conversa');
      }
    } catch (e) {
      debugPrint('❌ Erro ao deletar conversa: $e');
      rethrow;
    }
  }

  /// Atualiza o título de uma conversa
  Future<void> updateTitle(int conversationId, String newTitle) async {
    try {
      final response = await http.patch(
        Uri.parse('$baseUrl/$conversationId/title'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'title': newTitle}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        if (data['success'] != true) {
          throw Exception('Falha ao atualizar título');
        }
      } else if (response.statusCode == 404) {
        throw Exception('Conversa não encontrada');
      } else {
        final error = jsonDecode(response.body) as Map<String, dynamic>;
        throw Exception(error['detail'] ?? 'Erro ao atualizar título');
      }
    } catch (e) {
      debugPrint('❌ Erro ao atualizar título: $e');
      rethrow;
    }
  }
}

